pylint/plugins/hass_enforce_type_hints.py

```diff
@@ -510,38 +498,32 @@ class HassTypeHintChecker(BaseChecker):  # type: ignore[misc]
 
     def __init__(self, linter: PyLinter | None = None) -> None:
         super().__init__(linter)
-        self.current_package: str | None = None
-        self.module: str | None = None
+        self._function_matchers: list[TypeHintMatch] = []
 
     def visit_module(self, node: astroid.Module) -> None:
         """Called when a Module node is visited."""
-        self.module = node.name
-        if node.package:
-            self.current_package = node.name
-        else:
-            # Strip name of the current module
-            self.current_package = node.name[: node.name.rfind(".")]
+        self._function_matchers = []
+        for key, pattern in _MODULE_FILTERS.items():
+            if pattern.match(node.name) and (matches := _METHOD_MATCH.get(key)):
+                self._function_matchers.extend(matches)
 
     def visit_functiondef(self, node: astroid.FunctionDef) -> None:
         """Called when a FunctionDef node is visited."""
-        for match in _METHOD_MATCH:
-            self._visit_functiondef(node, match)
+        self._visit_functiondef(node)
 
     def visit_asyncfunctiondef(self, node: astroid.AsyncFunctionDef) -> None:
         """Called when an AsyncFunctionDef node is visited."""
-        for match in _METHOD_MATCH:
-            self._visit_functiondef(node, match)
+        self._visit_functiondef(node)
 
-    def _visit_functiondef(
-        self, node: astroid.FunctionDef, match: TypeHintMatch
-    ) -> None:
-        if node.name != match.function_name:
-            return
-        if node.is_method():
-            return
-        if not match.module_filter.match(self.module):
-            return
+    def _visit_functiondef(self, node: astroid.FunctionDef) -> None:
+        for match in self._function_matchers:
+            if node.name != match.function_name:
+                continue
+            if node.is_method():
+                continue
```

Comment:
```suggestion
            if node.name != match.function_name or node.is_method():
                continue
```

Reply:
I'll adjust that in a follow-up PR, together with the astroid import and adding the generic to Pattern

Reply:
I changed my mind and implemented this now (not in follow-up PR)